### PR TITLE
Add sbt-based bootstrap

### PR DESCRIPTION
**Pull request by smarter**
This PR might be the hardest I've worked on just because of all the deep and subtle issues in dotty it forced me to figure out (especially #1770, #1856, #1895), but is now finally ready! ~(Well, almost, the backend needs to updated with https://github.com/DarkDimius/scala/pull/4 first and we need to get https://github.com/lampepfl/dotty/pull/1894 in, but other than that it should be complete)~.

This adds two new project to the sbt build: dotty-library-bootstrapped and dotty-compiler-bootstrapped. These projects use the same source files as dotty-library and dotty-compiler but are compiled using dotty itself. The main usecase for this is that we can now run the JUnit tests (which are _not_ just a subset of the tests run by partest, for example the REPL tests are only run through JUnit) with a bootstrapped compiler:

```
$ sbt
> publishLocal # Non-bootstrapped dotty needs to be published first
> dotty-bootstrapped/test
```

But this also allows one to experiment with a bootstrapped dotty in a much nicer way. This also replaces the previous bootstrap mechanism used for partest: `partest` and `partest-only` will now go through `dotty-compiler-bootstrapped`. Since partest bootstrapping has been broken for several months now, I consider that a win (we could also keep the old mechanism around but only if someone volunteers to maintain it).

### CODE REVIEW

**Comment by smarter** [comment]
Review by @felixmulder for the sbt/tests related stuff and by @odersky for everything else :).

**Comment by smarter** [comment]
Every day is a new adventure, apparently compiling dotty (with scalac) is enough to fill up the code cache of the JVM now? http://dotty-ci.epfl.ch/lampepfl/dotty/530/2
```scala
CodeCache: size=131072Kb used=124845Kb max_used=125852Kb free=6226Kb
 bounds [0x00007f0cd0d0f000, 0x00007f0cd8d0f000, 0x00007f0cd8d0f000]
 total_blobs=27273 nmethods=26706 adapters=462
 compilation: disabled (not enough contiguous free space left)
OpenJDK 64-Bit Server VM warning: CodeCache is full. Compiler has been disabled.
OpenJDK 64-Bit Server VM warning: Try increasing the code cache size using -XX:ReservedCodeCacheSize=
```

**Comment by retronym** [comment]
> Every day is a new adventure, apparently compiling dotty (with scalac) is enough to fill up the code cache of the JVM now? http://dotty-ci.epfl.ch/lampepfl/dotty/530/2

We decided to handle bootstrapping outside of the SBT build in the Scala build for this reason. We use `sbt .... publishLocal; sbt -Dstarr.version=....` instead.

**Comment by smarter** [comment]
I think I can handle it by restricting sbt task parallelism when I bootstrap (right now it's happily compiling/publishing everything at the same time)

**Comment by smarter** [comment]
Oh, I think what's going is that it's compiling dotty twice at the same time, I thought sbt whole job was to not do that sort of things:
```scala
[info] Compiling 296 Scala sources and 1 Java source to /drone/src/github.com/lampepfl/dotty/compiler/target/scala-2.11/classes...
...
[info] Compiling 296 Scala sources and 1 Java source to /drone/src/github.com/lampepfl/dotty/compiler/target/scala-2.11/classes...
```

**Comment by DarkDimius** [comment]
DarkDimius/scala#4 was merged

File: project/Build.scala [hunk 1]
```diff
@@ -251,8 +242,8 @@ object DottyBuild extends Build {
 
       // enable verbose exception messages for JUnit
       testOptions in Test += Tests.Argument(
-        TestFrameworks.JUnit, "-a", "-v",
-        "--run-listener=dotty.tools.ContextEscapeDetector"
+        TestFrameworks.JUnit, "-a", "-v"/*,
+        "--run-listener=dotty.tools.ContextEscapeDetector"*/
```
Comment (felixmulder) [inline project/Build.scala]: Does this need to be disabled still?

File: project/Build.scala [hunk 2]
```diff
@@ -371,17 +418,36 @@ object DottyBuild extends Build {
...
+    settings(
+      // FIXME: this breaks the CI (OutOfMemory errors) somehow
+      // The non-bootstrapped compiler needs to be published first.
+      // update <<= update.dependsOn(Def.task {
+      //   // This crazy stuff is equivalent to `dotty/publishLocal`, if we just did
+      //   // `publishLocal in dotty` the aggregated projects would be ignored.
+      //   val s = state.value
+      //   val e = Project.extract(s)
+      //   e.runAggregated(publishLocal in dotty, s)
+      // })
```
Comment (felixmulder) [inline project/Build.scala]: comment in empty settings?

Comment (smarter) [inline project/Build.scala]: Yeah I should comment the whole thing

File: project/Build.scala (hunk 1 above)
Comment (smarter) [inline project/Build.scala]: Until we have a published new backend and https://github.com/lampepfl/dotty/pull/1894 is merged, yes. (I'll wait for these things to happen before merging this PR)

**Comment by smarter** [comment]
Rebased to master which contains https://github.com/lampepfl/dotty/pull/1912 which was the last blocker to get this in

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 3]
```diff
@@ -319,7 +319,17 @@ class Definitions {
...
+  // Dotty deviation: we cannot use a lazy val here because lazy vals in dotty
+  // will return "null" when called recursively, see #1856.
+  def DottyPredefModuleRef = {
+    if (_DottyPredefModuleRef == null) {
+      _DottyPredefModuleRef = ctx.requiredModuleRef("dotty.DottyPredef")
+      assert(_DottyPredefModuleRef != null)
+    }
+    _DottyPredefModuleRef
+  }
+  private[this] var _DottyPredefModuleRef: TermRef = _
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: Aside: Following @ichoran's example, I tend to use `myDottyPredefModuleRef` for private vars that have a getter `DottyPredefModuleRef`. But _... notation works as well.

File: compiler/test/dotty/partest/DPConsoleRunner.scala [hunk 4]
```diff
@@ -398,8 +398,13 @@ class DPTestRunner(testFile: File, suiteRunner: DPSuiteRunner) extends nest.Runn
   override def extraClasspath =
     suiteRunner.fileManager.asInstanceOf[DottyFileManager].extraJarList ::: super.extraClasspath
 
+
+  // FIXME: Dotty deviation: error if return type is emitted:
```
Comment (odersky) [inline compiler/test/dotty/partest/DPConsoleRunner.scala]: emitted -> omitted

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 3 above)
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala]: OK, let's standardize on my

File: compiler/test/dotty/partest/DPConsoleRunner.scala (hunk 4 above)
Comment (smarter) [inline compiler/test/dotty/partest/DPConsoleRunner.scala]: Fixed.

**Pull request merged**